src/server/frontend_wayland/wl_surface.cpp

```diff
@@ -499,3 +500,77 @@ auto mf::NullWlSurfaceRole::scene_surface() const -> std::optional<std::shared_p
 void mf::NullWlSurfaceRole::refresh_surface_data_now() {}
 void mf::NullWlSurfaceRole::commit(WlSurfaceState const& state) { surface->commit(state); }
 void mf::NullWlSurfaceRole::surface_destroyed() {}
+
+mf::DragWlSurfaceRole::DragWlSurfaceRole(WlSurface* surface)
+    : NullWlSurfaceRole(surface),
+      surface{surface}
+{
+    surface->set_role(this);
+    create_scene_surface();
+}
+
+mf::DragWlSurfaceRole::~DragWlSurfaceRole()
+{
+    if (surface)
+    {
+        surface.value().clear_role();
+
+        if (shared_scene_surface)
+        {
+            auto const& session = surface.value().session;
+            session->destroy_surface(shared_scene_surface);
```

Comment:
This is the wrong place to remove the scene surface; you should be implementing `surface_destroyed()` (which is called when the underlying `wl_surface` is destroyed) and removing the scene surface then.

This destructor is not called when the underlying `WlSurface` is destroyed.

Reply:
Hm. On second thoughts, this is *one of* the places that the scene surface needs to be removed, so that the scene surface is removed when the drag ends.

The *other* one is in a `surface_destroyed()` implementation, so that the scene surface is removed if the underlying `wl_surface` is destroyed.